actions/mongodb_explain_action.go

```diff
@@ -0,0 +1,94 @@
+// pmm-agent
+// Copyright 2019 Percona LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package actions
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/percona/percona-toolkit/src/go/mongolib/proto"
+	"github.com/percona/pmm/api/agentpb"
+	"github.com/pkg/errors"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+// MongodbExplainAction holds unexported fields and implements the Action interface
+type MongodbExplainAction struct {
+	id     string
+	params *agentpb.StartActionRequest_MongoDBExplainParams
+}
+
+// NewMongoDBExplainAction creates a MongoDB  EXPLAIN query Action.
+func NewMongoDBExplainAction(id string, params *agentpb.StartActionRequest_MongoDBExplainParams) Action {
+	return &MongodbExplainAction{
+		id:     id,
+		params: params,
+	}
+}
+
+// ID returns an Action ID.
+func (a *MongodbExplainAction) ID() string {
+	return a.id
+}
+
+// Type returns an Action type.
+func (a *MongodbExplainAction) Type() string {
+	return "mongodb-explain"
+}
+
+// Run runs an Action and returns output and error.
+func (a *MongodbExplainAction) Run(ctx context.Context) ([]byte, error) {
+	dsn := a.params.Dsn
+	client, err := mongo.Connect(ctx, options.Client().ApplyURI(dsn))
+	if err != nil {
+		return nil, err
+	}
+
+	// Check the connection
+	err = client.Ping(ctx, nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot connect to the database. Ping failed")
+	}
+	defer client.Disconnect(ctx) //nolint:errcheck
```

Comment:
Seems like that defer should be defined on line `61`, right after error check of `mongo.Connect`. As the documentation says that `mongo.Connect` starts some background goroutines for monitoring and `client.Disconnect` terminates suck goroutines.

Reply:
Nice catch. Thanks. Already updated